Bypass const_item_mutation if const's type has Drop impl

dtolnay opened the pull request:
Follow-up to #75573. This PR disables the const_item_mutation lint in cases that the const has a Drop impl which observes the mutation.

```rust
struct Log { msg: &'static str }
const LOG: Log = Log { msg: "" };
impl Drop for Log {
    fn drop(&mut self) { println!("{}", self.msg); }
}

LOG.msg = "wow";  // prints "wow"
```

r? @Aaron1011

dtolnay committed:
Bypass const_item_mutation if const's type has Drop impl

dtolnay committed:
Add justification of the destructor filter

dtolnay committed:
Simplify defid destructor check

dtolnay committed:
Add test of const item mutation with Drop impl

dtolnay committed:
Skip dropck::check_drop_impl in is_const_item_without_destructor

adt_destructor by default also validates the Drop impl using
dropck::check_drop_impl, which contains an expect_local(). This
leads to ICE in check_const_item_mutation if the const's type is
not a local type.

    thread 'rustc' panicked at 'DefId::expect_local: `DefId(5:4805 ~ alloc[d7e9]::vec::{impl#50})` isn't local', compiler/rustc_span/src/def_id.rs:174:43
    stack backtrace:
       0: rust_begin_unwind
       1: rustc_span::def_id::DefId::expect_local::{{closure}}
       2: rustc_typeck::check::dropck::check_drop_impl
       3: rustc_middle::ty::util::<impl rustc_middle::ty::context::TyCtxt>::calculate_dtor::{{closure}}
       4: rustc_middle::ty::trait_def::<impl rustc_middle::ty::context::TyCtxt>::for_each_relevant_impl
       5: rustc_middle::ty::util::<impl rustc_middle::ty::context::TyCtxt>::calculate_dtor
       6: rustc_typeck::check::adt_destructor
       7: rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::adt_destructor>::compute
       8: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl
       9: rustc_query_system::query::plumbing::get_query_impl
      10: rustc_mir::transform::check_const_item_mutation::ConstMutationChecker::is_const_item_without_destructor

Aaron1011 commented:
I'm not convinced that this is a good idea. The point of `const_item_mutation` is to warn on counterintuitive code which appears to have a side effect but actually doesn't. 

If anything, `LOG.msg = "wow"` seems even *more* counterintuitive than the 'typical' code that const_item_mutation warns against - it has a side effect that you wouldn't expect, and *doesn't* have the side effect that you *would* expect.

However, I think it would be reasonable to add a note to the warning message if the type has a destructor, so that the user is away that deleting the statement may cause unexpected changes in behavior.

dtolnay commented:
Let me clarify the reason I came across this. I am using assignments for a global static configuration of a library, as in e.g. `CFG.header_include_prefix = "path/to"`. It isn't so much about println-like side effects, rather the assignment actually being very much the intended effect. We just don't exposing the whole *MUTEX.lock().unwrap() rigmarole that would usually follow with a static config.

Here is a simplified setup to give you the flavor (actual implementation is somewhat different):

```rust
pub mod thelibrary {
    use lazy_static::lazy_static;
    use std::ptr;
    use std::sync::atomic::{AtomicPtr, Ordering};
    use std::sync::Mutex;

    pub struct Options {
        pub include_prefix: &'static str,
    }

    lazy_static! {
        static ref INCLUDE_PREFIX: Mutex<&'static str> = Mutex::new(env!("CARGO_CRATE_NAME"));
    }

    impl Options {
        fn current() -> Self {
            Options {
                include_prefix: *INCLUDE_PREFIX.lock().unwrap(),
            }
        }
    }

    use private::Cfg;
    pub const CFG: Cfg = Cfg {
        options: AtomicPtr::new(ptr::null_mut()),
    };
    mod private {
        use super::*;
        pub struct Cfg {
            pub(super) options: AtomicPtr<Options>,
        }
    }

    impl std::ops::Deref for Cfg {
        type Target = Options;
        fn deref(&self) -> &Self::Target {
            let options = Box::into_raw(Box::new(Options::current()));
            let prev = self
                .options
                .compare_and_swap(ptr::null_mut(), options, Ordering::Relaxed);
            if !prev.is_null() {
                // compare_and_swap did nothing.
                let _ = unsafe { Box::from_raw(options) };
                panic!();
            }
            unsafe { &*options }
        }
    }

    impl std::ops::DerefMut for Cfg {
        fn deref_mut(&mut self) -> &mut Self::Target {
            let options = self.options.get_mut();
            if !options.is_null() {
                panic!();
            }
            *options = Box::into_raw(Box::new(Options::current()));
            unsafe { &mut **options }
        }
    }

    impl Drop for Cfg {
        fn drop(&mut self) {
            let options = *self.options.get_mut();
            if let Some(options) = unsafe { options.as_mut() } {
                *INCLUDE_PREFIX.lock().unwrap() = options.include_prefix;
            }
        }
    }
}

use thelibrary::CFG;

fn main() {
    CFG.include_prefix = "path/to";

    println!("{:?}", CFG.include_prefix);
}
```

> If anything, ~~`LOG.msg = "wow"`~~`CFG.include_prefix = "path/to"` seems even *more* counterintuitive than the 'typical' code that const_item_mutation warns against - it has a side effect that you wouldn't expect, and *doesn't* have the side effect that you *would* expect.

I think the snippet above shows it has exactly the effect the person should expect. The write is observable to subsequent reads of the const.

Aaron1011 commented:
Wow, that's quite the API :smile: 

It looks like the lint is firing for the implicit `DerefMut` call, rather than for a direct assignment. In general, I think this is correct - writing to a field via a const item's `DerefMut` impl will almost never have the desired effect.

I'm worried that a blanket suppression of this lint for all types with drop glue will cause many false negatives. I think it would make more sense to have a per-item opt out, which would indicate that any mutation actually has some side effect.

@dtolnay Would something like `#[const_mutation_allowed] pub const CFG` fit your use case?

dtolnay commented:
> Wow, that's quite the API :smile: 

It has been beneficial for this library because it very deliberately triggers "okay I am writing to a global (though safely)" neurons, which `::set_whatever(...)` does not, in my experience, at least not nearly as strongly or reliably.

> Would something like `#[const_mutation_allowed] pub const CFG` fit your use case?

Right now that would be slightly less good for me because it would imply having to do rustc version detection for whether const_mutation_allowed is supported in the current compiler, but it would work, and I sympathize with the concern about false negatives.

As a new user-facing attribute I guess that would need to go through FCP (compiler team? lang team?). If so, and if it doesn't make the next beta cutoff which is happening next week, I would ask that we consider the approach in the PR in the interim until const_mutation_allowed is landed (and not feature gated).

Aaron1011 commented:
> As a new user-facing attribute I guess that would need to go through FCP (compiler team? lang team?). If so, and if it doesn't make the next beta cutoff which is happening next week, I would ask that we consider the approach in the PR in the interim until const_mutation_allowed is landed (and not feature gated).

That sounds fine to me. I'll work on an FCP.

dtolnay commented:
> I'm worried that a blanket suppression of this lint for all types with drop glue will cause many false negatives.

To clarify, as implemented in the PR this is *not* bypassing for all types with drop glue. It's only for types that specifically have their own Drop impl. Something with compiler-generated drop glue but not a Drop impl would still warn.

```rust
struct Log { s: String }
const LOG: Log = Log { s: String::new() };

fn main() {
    LOG.msg = "~~~".to_owned(); //~attempting to modify a `const` item
}
```

dtolnay committed:
Test a type with drop glue but no Drop impl

bugaevc commented:
Would a warning still get emitted in the `Vec` case?


```rust
const VEC: Vec<i32> = Vec::new();

fn main() {
    VEC.push(42);
    println!("{:?}", VEC);
}
```

dtolnay commented:
@bugaevc: Vec has a drop impl so it would not warn.

https://github.com/rust-lang/rust/blob/db7ee7cc0ddf46a52f53a8a141cd0747f829999a/library/alloc/src/vec.rs#L2631

bugaevc commented:
What I'm saying is it would be unfortunate to lose the warning in the `Vec` case; this is specifically the case where I've seen people get confused.

So perhaps there could be a better heuristic to detect cases like yours above, while also keeping a warning for simple cases that happen to have a `Drop` impl?

bors commented:
:umbrella: The latest upstream changes (presumably #77381) made this pull request unmergeable. Please resolve the merge conflicts.

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

dtolnay committed:
Add test of VEC.push on a const

dtolnay committed:
Warn on method call mutating const, even if it has destructor

dtolnay commented:
- Rebased to resolve conflict with #77324.

dtolnay commented:
@bugaevc: I updated the PR to still warn in the case of a `&mut self` parameter of a method attempting to mutate a const even if there is a destructor, such as in `VEC.push`, and added a test.

Aaron1011 commented on compiler/rustc_mir/src/transform/check_const_item_mutation.rs:
```diff
@@ -31,6 +31,28 @@ impl<'a, 'tcx> ConstMutationChecker<'a, 'tcx> {
+        let def_id = self.is_const_item(local)?;
+        let mut any_dtor = |_tcx, _def_id| Ok(());
+
+        // We avoid linting mutation of a const item if the const's type has a
```
Can you open an issue to track removing this (conditional on having a stable attribute to suppress the lint for a particular `const` item), and add a FIXME?

Aaron1011 commented:
@dtolnay: I wasn't able to come up with a good design for an attribute to suppress this. There are similar lints that I'd like to create/uplift in the future (e.g. const items with interior mutability), and it would be annoying to introduce a single-use attribute for each of them.

r=me with the comment addressed, so that this can be beta-backported.

dtolnay committed:
Fixme with link for re-enabling const mutation lint for Drop consts

dtolnay commented:
@bors r=Aaron1011

bors commented:
:pushpin: Commit 804d159c6276025e26ade90a6ef013bda33207c3 has been approved by `Aaron1011`

<!-- @bors r=Aaron1011 804d159c6276025e26ade90a6ef013bda33207c3 -->
<!-- homu: {"type":"Approved","sha":"804d159c6276025e26ade90a6ef013bda33207c3","approver":"Aaron1011"} -->
